### PR TITLE
18468 - EFT Credits / Model updates

### PR DESCRIPTION
**Pull request by ochiu**
*Issue #:*
https://github.com/bcgov/entity/issues/18468

*Description of changes:*
- Remove old EFT and WIRE cfs invoice creation job calls
- Add eft_credits table to track overpayment
- Add eft_short_names versioning table
- Implementation usage of eft_credits similar to existing cfs credits
- Refactor to separate CFS and EFT credits + feature flag
- EFT Credit test for applying credits fully, partially and remaining credit balance

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/services/payment_account.py [hunk 1]
```diff
@@ -556,6 +561,52 @@ def find_by_id(cls, account_id: int):
...
 
+    @classmethod
+    def get_eft_credit_balance(cls, account_id: int) -> Decimal:
+        """Calculate pay account eft balance by account id."""
+        result = db.session.query(func.sum(EFTCreditModel.remaining_amount).label('credit_balance')) \
+            .filter(EFTCreditModel.payment_account_id == account_id) \
+            .group_by(EFTCreditModel.payment_account_id) \
+            .one_or_none()
+
+        return Decimal(result.credit_balance) if result else 0
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:572]: str(result.credit_balance)

File: pay-api/src/pay_api/services/payment_account.py [hunk 2]
```diff
@@ -556,6 +561,52 @@ def find_by_id(cls, account_id: int):
...
+
+    @classmethod
+    def deduct_eft_credit(cls, account_id: int, invoice: Invoice) -> Decimal:
+        """Deduct EFT credit and update remaining credit records."""
+        # Get all eft credits
+        result: List[EFTCreditModel] = db.session.query(EFTCreditModel) \
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:578]: suggestion rename `result` -> `eft_credits`

I don't think you'd need the comment on 577 then

File: pay-api/src/pay_api/services/payment_account.py [hunk 3]
```diff
@@ -556,6 +561,52 @@ def find_by_id(cls, account_id: int):
...
+                eft_credit.save()
+
+                invoice.paid = invoice.total
+                invoice.invoice_status_code = InvoiceStatus.PAID.value
+                invoice.save()
+
+                break
+
+            # Credit covers partial invoice balance
+            invoice_balance = invoice_balance - eft_credit.remaining_amount
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:601]: `invoice_balance += -eft_credit.remaining_amount`

File: pay-api/src/pay_api/services/payment_account.py [hunk 4]
```diff
@@ -556,6 +561,52 @@ def find_by_id(cls, account_id: int):
...
+            invoice.paid = invoice.paid + eft_credit.remaining_amount
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:602]: `invoice.paid += eft_credit.remaining_amount`

File: pay-api/src/pay_api/services/payment_service.py [hunk 5]
```diff
@@ -180,23 +180,16 @@ def update_invoice(cls, invoice_id: int, payment_request: Tuple[Dict[str, Any]],
...
-                credit_balance = payment_account.credit - invoice_balance
-                invoice.paid = invoice.total
-                invoice.save()
-            elif (payment_account.credit or 0) <= invoice_balance:
-                invoice.paid = (invoice.paid or 0) + (payment_account.credit or 0)
-                invoice.save()
-
-            payment_account.credit = credit_balance
-            payment_account.save()
+
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_service.py:184]: I don't think this is needed, I don't think we call `update_invoice` for EFT in the frontend at all? 
I'll have to look a little closer

File: pay-api/tests/unit/api/test_payment_request.py [hunk 6]
```diff
@@ -881,6 +884,91 @@ def test_payment_request_online_banking_with_credit(session, client, jwt, app):
...
+    assert eft_credit_1.remaining_amount == 0
+    assert eft_credit_2.remaining_amount == 0
+
+    # Increase credit and test for left over balance
+    eft_credit_2.amount = 60
+    eft_credit_2.remaining_amount = 14.50
+    eft_credit_2.save()
+
+    # Apply credit to the previous partial invoice
+    client.patch(f'/api/v1/payment-requests/{invoice_id}?applyCredit=true', data=json.dumps({}), headers=headers)
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_payment_request.py:960]: I don't think this is needed, I'll have to look at the frontend that uses this

**Comment by seeker25** [comment]
I've remade the issue as a ticket instead of the epic, I've repointed the title and description at it.

File: pay-api/src/pay_api/services/payment_account.py (hunk 2 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/payment_account.py]: Done

File: pay-api/src/pay_api/services/payment_account.py (hunk 3 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/payment_account.py]: Updated

File: pay-api/src/pay_api/services/payment_account.py (hunk 4 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/payment_account.py]: Updated

File: pay-api/tests/unit/api/test_payment_request.py (hunk 6 above)
Comment (ochiu) [reply pay-api/tests/unit/api/test_payment_request.py]: Reverted, and moved the test to the payment account service

File: pay-api/src/pay_api/services/payment_account.py [hunk 7]
```diff
@@ -556,6 +560,51 @@ def find_by_id(cls, account_id: int):
...
+            .all()
+
+        # Calculate invoice balance
+        invoice_balance = invoice.total - (invoice.paid or 0)
+
+        # Deduct credits and apply to the invoice
+        for eft_credit in eft_credits:
+            if eft_credit.remaining_amount >= invoice_balance:
+                # Credit covers the full invoice balance
+                eft_credit.remaining_amount = eft_credit.remaining_amount - invoice_balance
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:589]: `-=` here?

**Review by seeker25 — APPROVED** [review]

File: pay-api/src/pay_api/services/payment_account.py (hunk 7 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/payment_account.py]: Good catch, fixed

**Pull request merged**